[UI-side compositing] Safari occasionally crashes when scrolling underneath `NSScrollerImpPair`

whsieh opened the pull request:
#### b2362a5d6b0dcee521bbc38c226ea2216e2422e7
<pre>
[UI-side compositing] Safari occasionally crashes when scrolling underneath `NSScrollerImpPair`
<a href="https://bugs.webkit.org/show_bug.cgi?id=254484">https://bugs.webkit.org/show_bug.cgi?id=254484</a>
rdar://107139674

Reviewed by Simon Fraser.

After the changes in 259592@main, we now update `NSScrollerImpPair` in response to various scrolling
-related events off of the scrolling thread, with a lock (`m_scrollerImpPairLock`) to prevent us
from concurrently updating the `NSScrollerImpPair`. However, this is actually insufficient to ensure
thread safety since `NSScrollerImpPair` internally schedules work on the main thread underneath
`-[NSScrollerImpPair endScrollGesture]`. As such, it&apos;s possible to end up calling methods on an
`NSTimer` that has already been destroyed.

Since we can&apos;t prevent `NSScrollerImpPair` from rescheduling its hide timer on the main thread, we
need to instead dispatch all of our other calls to update the imp pair on the main thread. To
achieve this, we introduce a new helper method on `ScrollerPairMac` to asynchronously dispatch to
the main thread from the scrolling thread with the `NSScrollerImpPair`, or call the block
immediately in the case where we&apos;re already on the main thread, and deploy this in all the places
where we currently update the `NSScrollerImpPair`. See below for more details.

I wasn&apos;t able to reproduce the crash in either a test, or manually in MiniBrowser/Safari, but I was
able to reproduce it and verify this fix by injecting a `sleep()` in `NSScrollerImpPair`, after the
hiding timer is deallocated but before the pointer is reset to nil.

* LayoutTests/tiled-drawing/scrolling/fast-scroll-div-latched-mainframe-with-handler.html:

Adjust a test to avoid flakiness after this change, since we now dispatch to the main thread before
updating the imp pair.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
(WebCore::ScrollerMac::attach):

Make this use the new getter for `scrollbarStyle()`.

(WebCore::ScrollerMac::setHostLayer):

Use another helper method here to avoid mutating `NSScrollerImpPair` from a background thread.

* Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:

Make `ScrollerPairMac` thread-safe ref-counted, so that we can safely create references to it
from the scrolling thread, for blocks that are dispatched to the main thread.

We also remove `m_scrollerImpPairLock`, since all access to the `NSScrollerImpPair` now happens on
the main thread.

(WebCore::ScrollerPairMac::create):
(WebCore::ScrollerPairMac::lastKnownMousePosition const):
(WebCore::ScrollerPairMac::scrollbarStyle const):
(WebCore::ScrollerPairMac::scrollerImpVertical):
(WebCore::ScrollerPairMac::scrollerImpPair): Deleted.
(WebCore::ScrollerPairMac::WTF_RETURNS_LOCK): Deleted.
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
(-[WebScrollerImpPairDelegateMac scrollerImpPair:updateScrollerStyleForNewRecommendedScrollerStyle:]):

Make this use the new `setScrollbarStyle` helpers, which internally dispatch onto the main thread
before attempting to access `NSScrollerImpPair`.

(WebCore::ScrollerPairMac::ScrollerPairMac):
(WebCore::ScrollerPairMac::init):

Drive-by adjustments: remove some unnecessary `WebCore::` namespacing, now that this entire class is
already in the WebCore namespace.

(WebCore::ScrollerPairMac::~ScrollerPairMac):

Also ensure that the underlying `NSScrollerImpPair` is also destroyed on the main thread.

(WebCore::ScrollerPairMac::handleWheelEventPhase):
(WebCore::ScrollerPairMac::viewWillStartLiveResize):
(WebCore::ScrollerPairMac::viewWillEndLiveResize):
(WebCore::ScrollerPairMac::contentsSizeChanged):
(WebCore::ScrollerPairMac::handleMouseEvent):
(WebCore::ScrollerPairMac::updateValues):

Deploy `ensureOnMainThreadWithProtectedThis` in various places where we update `NSScrollerImpPair`.

(WebCore::ScrollerPairMac::visibleSize const):
(WebCore::ScrollerPairMac::valuesForOrientation):
(WebCore::ScrollerPairMac::setVerticalScrollerImp):
(WebCore::ScrollerPairMac::setHorizontalScrollerImp):
(WebCore::ScrollerPairMac::setScrollbarStyle):
(WebCore::ScrollerPairMac::ensureOnMainThreadWithProtectedThis):

Add a helper here to make it easier to dispatch methods that mutate or update the `NSScrollerImpPair`
onto the main thread. If we&apos;re on the scrolling thread, this dispatches asynchronously on to the
main thread; otherwise, we&apos;ll run the block right away if we&apos;re already on the main thread.

* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.h:

Turn `m_scrollerPair` into a `Ref`, now that the class is ref-counted.

* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
(WebCore::ScrollingTreeScrollingNodeDelegateMac::ScrollingTreeScrollingNodeDelegateMac):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::~ScrollingTreeScrollingNodeDelegateMac):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateFromStateNode):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::handleWheelEvent):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateScrollbarPainters):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::initScrollbars):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateScrollbarLayers):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::handleWheelEventPhase):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::handleMouseEventForScrollbars):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::viewWillStartLiveResize):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::viewWillEndLiveResize):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::viewSizeDidChange):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::scrollbarStateForOrientation const):
* Source/WebCore/platform/mac/ScrollTypesMac.h: Added.
(WebCore::nsScrollerStyle):
(WebCore::scrollbarStyle):

Update all these call sites to use `-&gt;` instead of `.` when accessing `m_scrollerPair`.

Canonical link: <a href="https://commits.webkit.org/262194@main">https://commits.webkit.org/262194@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bc5580f805ab418722fc74222fec1eb625dd3cd0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/860 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/884 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/919 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1202 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/759 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/848 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/934 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/969 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/972 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/868 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/813 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/809 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1142 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/863 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/806 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/805 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/777 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/832 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/1823 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/817 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/779 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/763 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/817 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/201 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/830 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:
```diff
@@ -86,29 +93,36 @@ class ScrollerPairMac {
-    WebCore::ScrollingTreeScrollingNode& m_scrollingNode;
+    void ensureOnMainThreadWithImpPair(Function<void(NSScrollerImpPair *)>&&);
+
+    bool m_inLiveResize { false };
```
Nit; group with other 8-bit things to minimize padding.

smfr commented on Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:
```diff
@@ -86,29 +93,36 @@ class ScrollerPairMac {
+    std::optional<FloatPoint> m_lastScrollOffset;
 
-    mutable Lock m_scrollerImpPairLock;
+    mutable RecursiveLock m_scrollerImpPairLock;
```
Do we still need the lock at all?

whsieh replied:
Good point! I think we can just remove it, since:

- Initialization (`init()`) happens on the main thread.
- We can just cache the last scrollbar style and return that, in `ScrollerPairMac::scrollbarStyle()`.

That should simplify the code a lot, too.

smfr commented on Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
```diff
@@ -111,14 +111,23 @@ - (void)scrollerImpPair:(NSScrollerImpPair *)scrollerImpPair setContentAreaNeeds
-    [scrollerImpPair setScrollerStyle:newRecommendedScrollerStyle];
+    UNUSED_PARAM(scrollerImpPair);
+
+    switch (newRecommendedScrollerStyle) {
```
We should just write conversion functions so you don't have to switch. And it would be better to refactor so there's only one call to `_scrollerPair->setScrollbarStyle()`.

whsieh replied:
Sounds good! I'll be adding new helpers to go between the two types [per the comment below](https://github.com/WebKit/WebKit/pull/11984#discussion_r1149699758).

smfr commented on Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
```diff
@@ -145,27 +154,30 @@ - (void)scrollerImpPair:(NSScrollerImpPair *)scrollerImpPair updateScrollerStyle
-    default:
-        break;
-    }
+    ensureOnMainThreadWithImpPair([phase](NSScrollerImpPair *impPair) {
```
Do we need to retain the impPair? Or do we need to pass it at all, since `ensureOnMainThreadWithImpPair` has access to |this| on the main thread?

whsieh replied:
Ah, so passing it in here was just a convenience (though it's true that `this` is safe to access, so I can just use `m_scrollerImpPair` directly) — I will remove the `NSScrollerImpPair *` argument and just use `m_scrollerImpPair`.

smfr commented on Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
```diff
@@ -145,27 +154,30 @@ - (void)scrollerImpPair:(NSScrollerImpPair *)scrollerImpPair updateScrollerStyle
     [m_scrollerImpPairDelegate invalidate];
     [m_scrollerImpPair setDelegate:nil];
+
+    ensureOnMainThread([scrollerImpPair = std::exchange(m_scrollerImpPair, nil)] {
```
Is scrollerImpPair a `RetainPtr<>` here?

whsieh replied:
Yep!

smfr commented on Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
```diff
@@ -302,6 +323,60 @@ - (void)scrollerImpPair:(NSScrollerImpPair *)scrollerImpPair updateScrollerStyle
+{
+    ensureOnMainThreadWithImpPair([style](NSScrollerImpPair *impPair) {
+        switch (style) {
+        case ScrollbarStyle::AlwaysVisible:
```
Same comment about type conversion.

smfr commented on Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
```diff
@@ -302,6 +323,60 @@ - (void)scrollerImpPair:(NSScrollerImpPair *)scrollerImpPair updateScrollerStyle
+    // grab the lock before we either read or write from this property. However, we could refactor this logic
+    // to avoid the problem altogether, by caching the last known scroller style on `ScrollerPairMac` and
+    // just returning that.
+    switch ([m_scrollerImpPair scrollerStyle]) {
```
This could use type conversion helpers.

whsieh replied:
Makes sense — I'll add new helpers to go from `ScrollbarStyle` to `NSScrollerStyle` and back in a new `WebCore/platform/mac/ScrollTypesMac.h` header (since `ScrollTypes.h` and `ScrollerPairMac.h` are both included in pure C++ files, so I can't import the requisite AppKit headers that define `NSScrollerStyle` there).

whsieh committed:
[UI-side compositing] Safari occasionally crashes when scrolling underneath `NSScrollerImpPair`
https://bugs.webkit.org/show_bug.cgi?id=254484
rdar://107139674

Reviewed by Simon Fraser.

After the changes in 259592@main, we now update `NSScrollerImpPair` in response to various scrolling
-related events off of the scrolling thread, with a lock (`m_scrollerImpPairLock`) to prevent us
from concurrently updating the `NSScrollerImpPair`. However, this is actually insufficient to ensure
thread safety since `NSScrollerImpPair` internally schedules work on the main thread underneath
`-[NSScrollerImpPair endScrollGesture]`. As such, it's possible to end up calling methods on an
`NSTimer` that has already been destroyed.

Since we can't prevent `NSScrollerImpPair` from rescheduling its hide timer on the main thread, we
need to instead dispatch all of our other calls to update the imp pair on the main thread. To
achieve this, we introduce a new helper method on `ScrollerPairMac` to asynchronously dispatch to
the main thread from the scrolling thread with the `NSScrollerImpPair`, or call the block
immediately in the case where we're already on the main thread, and deploy this in all the places
where we currently update the `NSScrollerImpPair`. See below for more details.

I wasn't able to reproduce the crash in either a test, or manually in MiniBrowser/Safari, but I was
able to reproduce it and verify this fix by injecting a `sleep()` in `NSScrollerImpPair`, after the
hiding timer is deallocated but before the pointer is reset to nil.

* LayoutTests/tiled-drawing/scrolling/fast-scroll-div-latched-mainframe-with-handler.html:

Adjust a test to avoid flakiness after this change, since we now dispatch to the main thread before
updating the imp pair.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
(WebCore::ScrollerMac::attach):

Make this use the new getter for `scrollbarStyle()`.

(WebCore::ScrollerMac::setHostLayer):

Use another helper method here to avoid mutating `NSScrollerImpPair` from a background thread.

* Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:

Make `ScrollerPairMac` thread-safe ref-counted, so that we can safely create references to it
from the scrolling thread, for blocks that are dispatched to the main thread.

We also remove `m_scrollerImpPairLock`, since all access to the `NSScrollerImpPair` now happens on
the main thread.

(WebCore::ScrollerPairMac::create):
(WebCore::ScrollerPairMac::lastKnownMousePosition const):
(WebCore::ScrollerPairMac::scrollbarStyle const):
(WebCore::ScrollerPairMac::scrollerImpVertical):
(WebCore::ScrollerPairMac::scrollerImpPair): Deleted.
(WebCore::ScrollerPairMac::WTF_RETURNS_LOCK): Deleted.
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
(-[WebScrollerImpPairDelegateMac scrollerImpPair:updateScrollerStyleForNewRecommendedScrollerStyle:]):

Make this use the new `setScrollbarStyle` helpers, which internally dispatch onto the main thread
before attempting to access `NSScrollerImpPair`.

(WebCore::ScrollerPairMac::ScrollerPairMac):
(WebCore::ScrollerPairMac::init):

Drive-by adjustments: remove some unnecessary `WebCore::` namespacing, now that this entire class is
already in the WebCore namespace.

(WebCore::ScrollerPairMac::~ScrollerPairMac):

Also ensure that the underlying `NSScrollerImpPair` is also destroyed on the main thread.

(WebCore::ScrollerPairMac::handleWheelEventPhase):
(WebCore::ScrollerPairMac::viewWillStartLiveResize):
(WebCore::ScrollerPairMac::viewWillEndLiveResize):
(WebCore::ScrollerPairMac::contentsSizeChanged):
(WebCore::ScrollerPairMac::handleMouseEvent):
(WebCore::ScrollerPairMac::updateValues):

Deploy `ensureOnMainThreadWithProtectedThis` in various places where we update `NSScrollerImpPair`.

(WebCore::ScrollerPairMac::visibleSize const):
(WebCore::ScrollerPairMac::valuesForOrientation):
(WebCore::ScrollerPairMac::setVerticalScrollerImp):
(WebCore::ScrollerPairMac::setHorizontalScrollerImp):
(WebCore::ScrollerPairMac::setScrollbarStyle):
(WebCore::ScrollerPairMac::ensureOnMainThreadWithProtectedThis):

Add a helper here to make it easier to dispatch methods that mutate or update the `NSScrollerImpPair`
onto the main thread. If we're on the scrolling thread, this dispatches asynchronously on to the
main thread; otherwise, we'll run the block right away if we're already on the main thread.

* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.h:

Turn `m_scrollerPair` into a `Ref`, now that the class is ref-counted.

* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
(WebCore::ScrollingTreeScrollingNodeDelegateMac::ScrollingTreeScrollingNodeDelegateMac):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::~ScrollingTreeScrollingNodeDelegateMac):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateFromStateNode):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::handleWheelEvent):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateScrollbarPainters):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::initScrollbars):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateScrollbarLayers):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::handleWheelEventPhase):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::handleMouseEventForScrollbars):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::viewWillStartLiveResize):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::viewWillEndLiveResize):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::viewSizeDidChange):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::scrollbarStateForOrientation const):
* Source/WebCore/platform/mac/ScrollTypesMac.h: Added.
(WebCore::nsScrollerStyle):
(WebCore::scrollbarStyle):

Update all these call sites to use `->` instead of `.` when accessing `m_scrollerPair`.

Canonical link: https://commits.webkit.org/262194@main
